sermant-plugins/sermant-mq-grayscale/mq-grayscale-rocketmq-plugin/src/main/java/io/sermant/mq/grayscale/service/RocketMqConsumerGroupAutoCheck.java

```diff
@@ -0,0 +1,228 @@
+            GroupList groupList = mqClientApi.queryTopicConsumeByWho(brokerAddress, clientConfig.getTopic(),
+                    ROCKET_MQ_READ_TIMEOUT);
+            for (String group : groupList.getGroupList()) {
+                try {
+                    List<String> consumerIds = mqClientApi.getConsumerIdListByGroup(brokerAddress, group,
+                            ROCKET_MQ_READ_TIMEOUT);
+                    if (consumerIds.isEmpty()) {
+                        continue;
+                    }
+                    String grayTag = StringUtils.substringAfterLast(group, clientConfig.getConsumerGroup() + "_");
+                    if (!StringUtils.isEmpty(grayTag)) {
+                        grayTags.add(grayTag);
+                    }
+                } catch (RemotingConnectException | RemotingSendRequestException | RemotingTimeoutException
+                         | MQBrokerException | InterruptedException e) {
+                    LOGGER.warning(String.format(Locale.ENGLISH, "[auto-check] can not find ids in group: [%s].",
+                            group));
+                }
+            }
+            resetAutoCheckGrayTagItems(grayTags, clientConfig);
+        } catch (MQClientException | InterruptedException | RemotingTimeoutException | RemotingSendRequestException
+                 | RemotingConnectException | MQBrokerException e) {
+            LOGGER.log(Level.FINE, String.format(Locale.ENGLISH, "[auto-check] error, message: %s",
+                    e.getMessage()), e);
+        }
+    }
+
+    /**
+     * compare current query grayGroupTag with collected last time, reset autoCheckGrayTagItems
+     *
+     * @param grayTags grayTags
+     * @param clientConfig MqConsumerClientConfig
+     */
+    private static void resetAutoCheckGrayTagItems(Set<String> grayTags, RocketMqConsumerClientConfig clientConfig) {
+        String addrTopicGroupKey = RocketMqSubscriptionDataUtils.buildAddrTopicGroupKey(clientConfig.getTopic(),
+                clientConfig.getConsumerGroup(), clientConfig.getAddress());
+        if (grayTags.isEmpty()) {
+            if (LAST_TOPIC_GROUP_GRAY_TAG.containsKey(addrTopicGroupKey)) {
+                RocketMqSubscriptionDataUtils.resetAutoCheckGrayTagItems(new ArrayList<>(), clientConfig);
+                LAST_TOPIC_GROUP_GRAY_TAG.remove(addrTopicGroupKey);
+            }
+            return;
+        }
+        HashSet<String> currentGroups = new HashSet<>(grayTags);
+        Set<String> lastTags = LAST_TOPIC_GROUP_GRAY_TAG.get(addrTopicGroupKey);
+        if (LAST_TOPIC_GROUP_GRAY_TAG.containsKey(addrTopicGroupKey)) {
+            currentGroups.removeAll(lastTags);
+        }
+        List<GrayTagItem> grayTagItems = new ArrayList<>();
+        if (!currentGroups.isEmpty() || grayTags.size() != lastTags.size()) {
+            for (String grayTag : grayTags) {
+                LOGGER.log(Level.INFO, String.format(Locale.ENGLISH, "[auto-check] current find gray tag: [%s].",
```

Comment:
Logs are placed outside the loop and described at one time.

Reply:
fixed